### PR TITLE
Fix wrong shift in Git Blame display

### PR DESCRIPTION
**Pull request by adelecourt**
Fix of the shift
Add of a unit test to check the Visit method works

### CODE REVIEW

**Comment by ddurschlag** [comment]
It looks like you might've accidentally added a bunch of files from the plugin repo to the main repo.

**Comment by adelecourt** [comment]
It isn't an accident. The "plugin repo" is used by my new test and is necessary to test the Visit method. I have created a file in this repo in which I have written a few lines. Then I have made some commits and this is why I need to push the "plugin repo".

**Comment by ddurschlag** [comment]
Sure, but did you need to push the object database for the plugin repo into the main repo? How about your local git config, with the path on your hard drive where your git checkout is stored? If it's intentional, fine, but I can't think of a reason it would be useful.

**Comment by adelecourt** [comment]
I don't think I clearly understand what is the problem.
I just want to add a new test repo, with only 3 commits, to execute my test scenario. In the test, I check that the commit for a specific line is the right commit.
Can you explain me more precisely what you expect me to do ?

**Comment by KirillOsenkov** [comment]
First off, @adelecourt welcome and thanks for your contribution! Always good to have tests.

@ddurschlag I think the .git repo folder being added is a test repo to test the git blame functionality.

I do have to say that I'd be hesitant to check in a bunch of low-level git objects inside our git repo (regardless whether Git actually supports this scenario which I hope it does).

Instead I'm wondering if your GitBlameVistorTests could create a new repo from scratch using git init and then make a couple test commits. I'd write a InitRepo.cmd with a few command-line git commands to initialize the repo to the state you want. Then use Process.Start() to call that InitRepo.cmd to make sure the test repo you expect is created in the directory you expect.

My advice: keep it simple and try to minimize the number of commits and files in the test repo, as well as file contents, just enough to test your scenario.

I hope this sounds reasonable? Thanks again!

File: src/GitGlyph/GitBlameVisitor.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ public GitBlameVisitor(Repository r, ILog logger)
         public string Visit(string text, IReadOnlyDictionary<string, string> context)
         {
             var path = System.IO.Path.GetFullPath(context[ContextKeys.FilePath]);
-            var blame = GetBlame(path)?.FirstOrDefault(bh => bh.FinalStartLineNumber.ToString() == context[ContextKeys.LineNumber]);
+            var blame = GetBlame(path)?.FirstOrDefault(bh => bh.FinalStartLineNumber == Int32.Parse(context[ContextKeys.LineNumber]) - 1);
```
Comment (KirillOsenkov) [inline src/GitGlyph/GitBlameVisitor.cs]: While you're here, I'd extract `Int32.Parse(context[ContextKeys.LineNumber]) - 1` into a separate local variable to make this line less complicated. Also not sure if perhaps you need to maybe use TryParse?

File: src/HtmlGenerator.Tests/GitBlameVisitorTests.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+using System.IO;
+using GitGlyph;
+using LibGit2Sharp;
+using Microsoft.SourceBrowser.HtmlGenerator.Utilities;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace HtmlGenerator.Tests
+{
+    [TestClass]
+    public class GitBlameVistorTests
```
Comment (KirillOsenkov) [inline src/HtmlGenerator.Tests/GitBlameVisitorTests.cs]: Vistor is misspelled, should be Visitor

**Comment by KirillOsenkov** [comment]
Actually if you're using LibGit2Sharp it's even easier to initialize a new repo from scratch from the C# code, no need to even have a `InitRepo.cmd`.

**Comment by ddurschlag** [comment]
Sorry, I misunderstood the checked-in repo. While I agree with Kirill that using libgit2sharp to create the repo is cleaner, the current set-up now makes sense. If you do end up checking in these files, I might consider a different name that makes clear it's test data, rather than a real repo, and removing anything unnecessary (e.g. the various hooks and paths that confused me).

**Comment by KirillOsenkov** [comment]
We have a special folder for test data:
\TestCode\

But I'd still prefer creating the repo on-the-fly using either a command batch script or LibGit2Sharp APIs :)

**Comment by adelecourt** [comment]
All your remarks have been taken in account !

File: src/HtmlGenerator.Tests/GitBlameVisitorTests.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        /// <summary>
+        /// Write content in the file Test.txt, stage this file and finally commit this change to the repository.
+        /// </summary>
+        /// <param name="content">Line written in the the file</param>
+        /// <param name="author">Signature of the author</param>
+        /// <param name="committer">Signature of the committer</param>
+        /// <param name="repo">Repository to commit to</param>
+        /// <param name="commitMessage">Message of the commit</param>
+        /// <returns>Commit of the change</returns>
+        private Commit writeAndCommit(String content, Signature author, Signature committer, Repository repo, String commitMessage)
```
Comment (KirillOsenkov) [inline src/HtmlGenerator.Tests/GitBlameVisitorTests.cs]: Method names always start with an uppercase: WriteAndCommit

**Comment by KirillOsenkov** [comment]
Thank you! Much much better :)

**Comment by ddurschlag** [comment]
Huge improvement in changed filecount!

Possible future improvement: consider using setup/teardown methods (https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.testtools.unittesting.testinitializeattribute.aspx) for setting up and tearing down the repo, to keep test(s) themselves short and meaningful.

**Pull request merged**